xonsh/environ.py

```diff
@@ -7,62 +7,188 @@
 import builtins
 import platform
 import subprocess
+from datetime import datetime
+from functools import partial
 from warnings import warn
 
 from xonsh.tools import TERM_COLORS
 
-def current_branch(cwd=None):
-    """Gets the branch for a current working directory. Returns None
-    if the cwd is not a repository.  This currently only works for git, 
-    bust should be extended in the future.
-    """
-    branch = None
-    cwd = os.getcwd() if cwd is None else cwd
+class PromptFormatter(dict):
```

Comment:
> Why is that?

Copying dict subclasses can have strange behaviour (http://code.activestate.com/lists/python-dev/135141/, https://hg.python.org/cpython/file/22a0c925a7c2/Objects/dictobject.c#l2003). MutableMapping is more general.  And in my experience, using ABCs has always been much safer than subclassing the built-in types.  I'd rather us use MutableMapping than ever have to go down this non-obvious debugging rabbit hole again.

Another good reason is consistency.  Other places in the xonsh code base use MutableMapping, and not dict.

> Inheriting from dict directly is less code to write and maintain, less objects to create (no need for a separate private dict to actually store the items), and somewhat faster methods (since the methods we don't override are implemented in C).

These things are not hard and only around 20 lines of code to implement.  Plus it gives us flexibility and safety.

I hope that this helps explain it.